contracts/single-collateral-factory/SingleCollateralFactory.sol

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: bsl-1.1
+
+/*
+  Copyright 2022 Unit Protocol: Artem Zakharov (az@unit.xyz).
+*/
+pragma solidity ^0.7.1;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./interfaces/IFactory.sol";
+import "./interfaces/IStageDeployer.sol";
+
+import "../USDP.sol";
+import "../VaultParameters.sol";
+import "../Vault.sol";
+
+contract SingleCollateralFactory is Ownable, IFactory {
+
+    IStageDeployer public immutable usdpAndVaultDeployer;
+    IStageDeployer public immutable registriesDeployer;
+    IStageDeployer public immutable parametersDeployer;
+    IStageDeployer public immutable managersDeployer;
+    IStageDeployer public immutable oracleAndHelpersDeployer;
+
+    mapping (uint => Deploy) deploys;
+    uint public deploysCount;
+
+    bool public deployOnlyForOwner = true;
+
+    constructor(
+        IStageDeployer usdpAndVaultDeployer_,
+        IStageDeployer registriesDeployer_,
+        IStageDeployer parametersDeployer_,
+        IStageDeployer managersDeployer_,
+        IStageDeployer oracleAndHelpersDeployer_
+    ) {
+        usdpAndVaultDeployer = usdpAndVaultDeployer_;
+        usdpAndVaultDeployer_.setManager(address(this));
+
+        registriesDeployer = registriesDeployer_;
+        registriesDeployer_.setManager(address(this));
+
+        parametersDeployer = parametersDeployer_;
+        parametersDeployer_.setManager(address(this));
+
+        managersDeployer = managersDeployer_;
+        managersDeployer_.setManager(address(this));
+
+        oracleAndHelpersDeployer = oracleAndHelpersDeployer_;
+        oracleAndHelpersDeployer_.setManager(address(this));
+    }
+
+    function getDeploy(uint deployId_) public view returns (Deploy memory) {
+        return deploys[deployId_];
+    }
+
+    function setDeployOnlyForOwner(bool value_) public onlyOwner {
+        deployOnlyForOwner = value_;
+    }
+
+    function initDeploy(
+        DeploySettings memory deploySettings_
+    ) public {
+        require(!deployOnlyForOwner || owner() == msg.sender, "FACTORY: UNAUTHORIZED");
+
+        Deploy storage deploy = deploys[deploysCount];
+        deploy.deploySettings = deploySettings_;
+
+        emit DeployInitialized(deploysCount);
+
+        deploysCount++;
+    }
+
+    function continueDeploy(uint deployId_) public {
+        Deploy storage deploy = deploys[deployId_];
+        require(deploy.deploySettings.collateral != address(0), "FACTORY: NOT_STARTED");
+        require(!deploy.isFinished, "FACTORY: FINISHED");
+
+        uint8 stage = deploy.stage;
+
+        if (stage == 0) {
+            address[] memory result = usdpAndVaultDeployer.deployStage(deploy);
+            deploy.usdp = result[0];
+            deploy.vaultParameters = result[1];
+            deploy.vault = result[2];
```

Comment:
I'd emit `deploy.vault` in an event and maybe even record in a mapping (`vault => deployId`) for better integration and on-chain verification.